fly/src/sorts.rs

```diff
@@ -387,94 +458,224 @@ impl Context<'_> {
+                }
+                self.sort_check_term_bool(&mut proof.assert.x)
+                    .map_err(|x| (x, Some(proof.assert.span)))?
             }
-            self.check_sort_exists_or_empty(&binder.sort)?;
-            let sort = if binder.sort == Sort::Id("".to_owned()) {
-                let var = self.vars.new_key(OptionSort(None));
-                binder.sort = Sort::Id(format!("var {}", var.0));
-                NamedSort::Unknown(var)
-            } else {
-                NamedSort::Known(vec![], binder.sort.clone())
-            };
-            // Now add it to the context, allowing it to shadow bindings from
-            // any outer scopes.
-            self.add_name(binder.name.clone(), sort, ShadowingConstraint::Allow)?;
         }
         Ok(())
     }
 
-    // unify a sort and a sort variable, or return an error
-    fn unify_var_value(&mut self, value: &Sort, var: &AbstractSort) -> Result<(), SortError> {
-        match var {
-            AbstractSort::Known(v) if value == v => Ok(()),
-            AbstractSort::Known(v) => Err(SortError::ExpectedButFoundSorts {
-                expected: value.clone(),
-                found: v.clone(),
-            }),
-            AbstractSort::Unknown(v) => self
-                .vars
-                .unify_var_value(*v, OptionSort(Some(value.clone()))),
+    fn sort_check_statements(
+        &mut self,
+        statements: &mut [ThmStmt],
+    ) -> Result<(), (SortError, Option<Span>)> {
+        for statement in statements {
+            self.sort_check_statement(statement)?
         }
+        Ok(())
     }
 
-    // unify two sort variables, or return an error
-    fn unify_var_var(&mut self, a: &AbstractSort, b: &AbstractSort) -> Result<(), SortError> {
-        match (a, b) {
-            (AbstractSort::Known(a), AbstractSort::Known(b)) if a == b => Ok(()),
-            (AbstractSort::Known(a), AbstractSort::Known(b)) => {
-                Err(SortError::UnificationFail(a.clone(), b.clone()))
+    fn get(&self, name: &str) -> Option<&RelationOrIndividual> {
+        self.bound_names.get(name)
+    }
+
+    fn get_individual(&self, name: &str) -> Result<MaybeUnknownSort, SortError> {
+        match self.get(name) {
+            Some(RelationOrIndividual::Relation(_, _)) => Err(SortError::Uncalled(name.to_owned())),
+            Some(RelationOrIndividual::Individual(ret)) => Ok(ret.clone()),
+            None => Err(SortError::UnknownVariable(name.to_owned())),
+        }
+    }
+
+    fn get_relation(&self, name: &str) -> Result<(Vec<Sort>, Sort), SortError> {
+        match self.get(name).cloned() {
+            Some(RelationOrIndividual::Individual(_)) => {
+                Err(SortError::Uncallable(name.to_owned()))
+            }
+            Some(RelationOrIndividual::Relation(args, ret)) => Ok((args, ret)),
+            None => Err(SortError::UnknownFunction(name.to_owned())),
+        }
+    }
+}
+
+struct InternalContext<'a, 'b> {
```

Comment:
Can these lifetime variables be the same now that scope is owned?

Reply:
yes, good call